server/src/main/java/org/elasticsearch/common/util/AbstractBigByteArray.java

```diff
@@ -68,22 +79,40 @@ protected static void fillBySelfCopy(byte[] page, int fromBytes, int toBytes, in
     /**
      * Bulk copies array to paged array
      */
-    protected void set(long index, byte[] buf, int offset, int len, byte[][] pages, int shift) {
+    protected void set(long index, byte[] buf, int offset, int len, int shift) {
         assert index + len <= size();
         int pageIndex = pageIndex(index);
         final int indexInPage = indexInPage(index);
         if (indexInPage + len <= pageSize()) {
-            System.arraycopy(buf, offset << shift, pages[pageIndex], indexInPage << shift, len << shift);
+            System.arraycopy(buf, offset << shift, getPageForWriting(pageIndex), indexInPage << shift, len << shift);
         } else {
             int copyLen = pageSize() - indexInPage;
-            System.arraycopy(buf, offset << shift, pages[pageIndex], indexInPage, copyLen << shift);
+            System.arraycopy(buf, offset << shift, getPageForWriting(pageIndex), indexInPage, copyLen << shift);
             do {
                 ++pageIndex;
                 offset += copyLen;
                 len -= copyLen;
                 copyLen = Math.min(len, pageSize());
-                System.arraycopy(buf, offset << shift, pages[pageIndex], 0, copyLen << shift);
+                System.arraycopy(buf, offset << shift, getPageForWriting(pageIndex), 0, copyLen << shift);
             } while (len > copyLen);
         }
     }
+
+    protected byte[] getPageForWriting(int pageIndex) {
+        byte[] foundPage = pages[pageIndex];
+        if (foundPage == ZERO_PAGE) {
```

Comment:
This is the new overhead we add to the `set` path relative to not having this. It seems to me, that it's a trivial overhead, especially when considering that we're fine with lots of redundant reads in `set` today. Plus, I've never seen `set` hot in profiling.